File: egp/src/egp/pf_gpu.py
"""GPU implementation of the fully adapted particle filter, via WebGPU.

Runs the WGSL compute shader in ``pf.wgsl`` — the same shader the companion
webapp executes in the browser — through `wgpu-py <https://github.com/pygfx/
wgpu-py>`_, which reaches any Vulkan/Metal/D3D12 device, including integrated
GPUs.  Same algorithm and diagnostics as :mod:`.pf`: one workgroup of 256
threads runs one replicate (the phases of a step are sequential, so a single
replicate can only use one workgroup's worth of parallelism), and a batch of
replicates runs as concurrent workgroups of one dispatch — which is where the
device gets filled and the speedup comes from.

The shader works in f32: log Phi through an Abramowitz-Stegun erf and the
Laplace continued fraction for the Mills ratio; its inverse seeded by Acklam's
branches with the tails computed from log p directly (where exp would cancel)
and polished by Newton.  Weight errors are ~1e-4 relative and draw errors
~1e-2 at 5 sigma — far below the estimator's statistical resolution; the
per-replicate estimates agree with :func:`.pf.particle_filter_loglik` to well
within one Monte-Carlo standard error, with identical ESS.  Per-step
log-likelihoods are read back as f32 and accumulated in f64 here, where the
lost-lock diagnostic is evaluated exactly as in the CPU filter.

``wgpu`` is an optional dependency: ``pip install egp[gpu]``.
"""

from __future__ import annotations

from pathlib import Path

import numpy as np

from .pf import DEFAULT_COLLAPSE_MARGIN, FilterResult, collapse_threshold

_SHADER_PATH = Path(__file__).with_name("pf.wgsl")

_device_cache: list = []  # [(device, pipeline)] once created


def gpu_available() -> bool:
    """True when wgpu is installed and a WebGPU adapter answers."""
    try:
        import wgpu
    except ImportError:
        return False
    try:
        return wgpu.gpu.request_adapter_sync(power_preference="high-performance") is not None
    except Exception:
        return False


def gpu_description() -> str:
    """Human-readable name of the adapter the filter would run on."""
    import wgpu

    info = wgpu.gpu.request_adapter_sync(power_preference="high-performance").info
    return f"{info.get('device', '?')} ({info.get('backend_type', '?')})"


def _get_device():
    """The process-wide device and compiled pipeline (compilation is ~100 ms)."""
    if _device_cache:
        return _device_cache[0]
    try:
        import wgpu
    except ImportError as exc:  # pragma: no cover - depends on environment
        raise RuntimeError(
            'the GPU engine needs the wgpu package: pip install "egp[gpu]"'
        ) from exc
    adapter = wgpu.gpu.request_adapter_sync(power_preference="high-performance")
    if adapter is None:  # pragma: no cover - depends on environment
        raise RuntimeError("no WebGPU adapter found; use the CPU engine")
    device = adapter.request_device_sync()
    module = device.create_shader_module(code=_SHADER_PATH.read_text())
    pipeline = device.create_compute_pipeline(
        layout="auto", compute={"module": module, "entry_point": "main"}
    )
    _device_cache.append((device, pipeline))
    return _device_cache[0]


def max_batch(n_particles: int, n_taps: int, n: int) -> int:
    """Replicates one dispatch can hold within the device's buffer limits."""
    device, _ = _get_device()
    limit = device.limits["max-storage-buffer-binding-size"]
    ring_len = max(n_taps - 1, 1)
    per_rep = max(n_particles * ring_len * 4, n_particles * 16, n * 8)
    fit = limit // per_rep
    if fit < 1:
        raise RuntimeError(
            f"one replicate needs {per_rep / 1e6:.0f} MB of GPU buffer, over the "
            f"device limit of {limit / 1e6:.0f} MB; reduce n_particles or use the CPU engine"
        )
    return int(fit)


def particle_filter_loglik_gpu(
    ys,
    taps,
    delta: float,
    n_particles: int,
    seeds,
    collapse_margin: float = DEFAULT_COLLAPSE_MARGIN,
    burn_in: int = 0,
    progress=None,
) -> list[FilterResult]:
    """Run one GPU batch: one particle filter per row of ``ys``, concurrently.

    Parameters mirror :func:`.pf.particle_filter_loglik` except that ``ys`` is
    a list/array of R equal-length symbol sequences and ``seeds`` is one u32
    per replicate for the shader's counter-based RNG (the observed sequences
    come from the caller, so the engines can be fed identical data).
    ``progress(done, total)`` is called as the step sweep advances.
    """
    device, pipeline = _get_device()
    import wgpu

    y_batch = np.ascontiguousarray(ys, dtype=np.int32)
    if y_batch.ndim != 2:
        raise ValueError("ys must be R sequences of equal length")
    R, n = y_batch.shape
    taps = np.asarray(taps, dtype=float).ravel()
    L = taps.size
    ring_len = L - 1
    h0 = float(taps[0])
    if ring_len < 1:
        raise ValueError("the GPU path needs L >= 2; the white process is exact on the CPU")
    if not h0 > 0:
        raise ValueError("taps[0] must be positive (use the minimum-phase factorization)")
    seeds = np.ascontiguousarray(seeds, dtype=np.uint32)
    if seeds.size != R:
        raise ValueError("one seed per replicate required")
    N = int(n_particles)
    if N < 2:
        raise ValueError("n_particles must be >= 2")
    burn_in = int(np.clip(burn_in, 0, max(0, n - 1)))
    if R > max_batch(N, L, n):
        raise RuntimeError(f"batch of {R} replicates exceeds GPU buffer limits")

    usage = wgpu.BufferUsage
    storage = usage.STORAGE | usage.COPY_DST

    def buf(size, u=storage):
        return device.create_buffer(size=max(int(size), 16), usage=u)

    buffers = {
        "uniform": buf(48, usage.UNIFORM | usage.COPY_DST),
        "ringA": buf(R * N * ring_len * 4),
        "ringB": buf(R * N * ring_len * 4),
        "ys": buf(R * n * 4),
        "stepOut": buf(R * n * 8, storage | usage.COPY_SRC),
        "scratch": buf(R * N * 16),
        "cum": buf(R * N * 4),
        "taps": buf(L * 4),
        "repSeeds": buf(R * 4),
    }
    order = ["uniform", "ringA", "ringB", "ys", "stepOut", "scratch", "cum", "taps", "repSeeds"]
    try:
        device.queue.write_buffer(buffers["ys"], 0, y_batch)
        device.queue.write_buffer(buffers["taps"], 0, taps.astype(np.float32))
        device.queue.write_buffer(buffers["repSeeds"], 0, seeds)
        bind_group = device.create_bind_group(
            layout=pipeline.get_bind_group_layout(0),
            entries=[
                {
                    "binding": i,
                    "resource": {"buffer": buffers[k], "offset": 0, "size": buffers[k].size},
                }
                for i, k in enumerate(order)
            ],
        )

        uniform = np.zeros(12, dtype=np.uint32)
        f32 = uniform.view(np.float32)
        uniform[0], uniform[1], uniform[2] = N, ring_len, n
        f32[8], f32[9], f32[10], f32[11] = h0, delta / h0, delta, np.log(N)

        # Steps per submit, sized to keep one dispatch well under any device
        # timeout; syncing (a tiny readback) every few chunks bounds how much
        # work is in flight and gives the progress callback real meaning.
        chunk = max(8, min(512, int(3e7 / (N * (ring_len + 30)))))
        chunks = range(0, n, chunk)
        sync_every = max(1, len(chunks) // 50)
        for ci, start in enumerate(chunks):
            uniform[3], uniform[4] = start, min(chunk, n - start)
            device.queue.write_buffer(buffers["uniform"], 0, uniform)
            encoder = device.create_command_encoder()
            comp = encoder.begin_compute_pass()
            comp.set_pipeline(pipeline)
            comp.set_bind_group(0, bind_group)
            comp.dispatch_workgroups(R)
            comp.end()
            device.queue.submit([encoder.finish()])
            if ci % sync_every == sync_every - 1:
                device.queue.read_buffer(buffers["stepOut"], 0, size=8)  # sync point
                if progress is not None:
                    progress(min(start + chunk, n), n)

        raw = device.queue.read_buffer(buffers["stepOut"])
        if progress is not None:
            progress(n, n)
    finally:
        for b in buffers.values():
            b.destroy()

    # Per-step values accumulate in f64 here, where the lost-lock diagnostic
    # is evaluated exactly as in the CPU filter.
    steps = np.frombuffer(raw, dtype=np.float32).reshape(R, n, 2).astype(np.float64)
    limit = collapse_threshold(delta, h0, collapse_margin)
    results = []
    for r in range(R):
        loglik = steps[r, :, 0]
        ess = steps[r, :, 1]
        lost = loglik < limit
        results.append(
            FilterResult(
                loglik_nats=float(loglik.sum()),
                loglik_counted=float(loglik[burn_in:].sum()),
                n_steps=n,
                burn_in=burn_in,
                mean_ess=float(ess.mean()),
                min_ess=float(ess.min()),
                n_lost_lock=int(lost.sum()),
                first_lost_step=int(np.argmax(lost)) if lost.any() else -1,
                collapse_threshold=limit,
            )
        )
    return results

File: webapp/src/egp/pfGpu.ts
/**
 * WebGPU implementation of the fully adapted particle filter — the optional
 * fast path beside the scalar pf.ts, same algorithm, same diagnostics.
 *
 * Design: one workgroup of 256 threads runs one replicate, looping over the
 * step sequence inside the shader with workgroup barriers between the phases
 * of a step (weights → max/scan reductions → systematic resample by binary
 * search → truncated-normal propagation). Several replicates therefore run
 * concurrently as independent workgroups of one dispatch, which is where the
 * speedup comes from — the per-step phases are sequential, so a single
 * replicate can only use one workgroup's worth of parallelism.
 *
 * WGSL is f32-only, so the log-space normal functions are f32 rebuilds, not
 * ports: Φ through the Abramowitz–Stegun erf (1.5e-7 absolute) above
 * x = −1.5 and the Laplace continued fraction for the Mills ratio below it;
 * Φ⁻¹(eʸ) seeded by Acklam's branches — the tails computed from log p and
 * −log p directly, where exp would cancel — and polished by Newton on log Φ,
 * skipped above x = 5 where f32 log Φ saturates to 0 and would corrupt the
 * seed. Weight errors land near 1e-4 relative and draw errors near 1e-2 at
 * 5σ, both far below the estimator's statistical resolution; per-step
 * log-likelihoods are read back as f32 and accumulated in f64 on the CPU,
 * where the lost-lock diagnostic is also evaluated, exactly as in pf.ts.
 *
 * Ancestor gathers shift the innovation window as they copy, so the tap
 * vector never rolls; ring buffers ping-pong on step parity. The RNG is a
 * counter-based PCG hash of (seed, stream, counter) — deterministic for a
 * given seed regardless of GPU scheduling.
 */
import { collapseThreshold, DEFAULT_COLLAPSE_MARGIN, type FilterResult } from './pf'
// The shader is shared verbatim with the Python package (egp/pf_gpu.py):
// one WGSL source, two hosts.
import SHADER from '../../../egp/src/egp/pf.wgsl?raw'

/** Concurrent replicates per dispatch; more gives diminishing returns for a
 * live readout and multiplies buffer sizes. */
const MAX_REPS = 8

export interface GpuRunOptions {
  taps: Float64Array
  delta: number
  /** One quantized sequence per replicate, all the same length. */
  ys: Int32Array[]
  nParticles: number
  /** One RNG seed per replicate for the filter's internal draws. */
  repSeeds: Uint32Array
  onProgress?: (done: number, total: number) => void
}

export class GpuParticleFilter {
  private constructor(
    private device: GPUDevice,
    private pipeline: GPUComputePipeline,
  ) {}

  static supported(): boolean {
    return typeof navigator !== 'undefined' && !!navigator.gpu
  }

  static async create(): Promise<GpuParticleFilter> {
    if (!GpuParticleFilter.supported()) throw new Error('WebGPU is not available here')
    const adapter = await navigator.gpu!.requestAdapter()
    if (!adapter) throw new Error('no WebGPU adapter available')
    const device = await adapter.requestDevice()
    const module = device.createShaderModule({ code: SHADER })
    const pipeline = await device.createComputePipelineAsync({
      layout: 'auto',
      compute: { module, entryPoint: 'main' },
    })
    return new GpuParticleFilter(device, pipeline)
  }

  destroy(): void {
    this.device.destroy()
  }

  /** Replicates one dispatch can run concurrently within buffer limits. */
  maxReps(nParticles: number, ringLen: number, nSteps: number): number {
    const limit = this.device.limits.maxStorageBufferBindingSize
    const perRep = Math.max(nParticles * ringLen * 4, nParticles * 16, nSteps * 8)
    return Math.max(1, Math.min(MAX_REPS, Math.floor(limit / Math.max(perRep, 1))))
  }

  /** Run one batch of replicates; resolves to one FilterResult per replicate. */
  async run(opts: GpuRunOptions): Promise<FilterResult[]> {
    const { taps, delta, ys, nParticles, repSeeds, onProgress } = opts
    const R = ys.length
    const n = ys[0].length
    const N = nParticles | 0
    const L = taps.length
    const ringLen = L - 1
    const h0 = taps[0]
    if (ringLen < 1) throw new Error('GPU path needs L >= 2 (white noise is exact on the CPU)')
    if (R < 1 || repSeeds.length !== R) throw new Error('one seed per replicate required')
    const { device, pipeline } = this

    const storage = GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_DST
    const ringBytes = R * N * ringLen * 4
    const buffers = {
      uniform: device.createBuffer({ size: 48, usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST }),
      ringA: device.createBuffer({ size: ringBytes, usage: storage }),
      ringB: device.createBuffer({ size: ringBytes, usage: storage }),
      ys: device.createBuffer({ size: R * n * 4, usage: storage }),
      stepOut: device.createBuffer({ size: R * n * 8, usage: storage | GPUBufferUsage.COPY_SRC }),
      scratch: device.createBuffer({ size: R * N * 16, usage: storage }),
      cum: device.createBuffer({ size: R * N * 4, usage: storage }),
      taps: device.createBuffer({ size: L * 4, usage: storage }),
      repSeeds: device.createBuffer({ size: R * 4, usage: storage }),
      readback: device.createBuffer({ size: R * n * 8, usage: GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ }),
    }
    try {
      const yAll = new Int32Array(R * n)
      ys.forEach((y, r) => yAll.set(y, r * n))
      device.queue.writeBuffer(buffers.ys, 0, yAll)
      device.queue.writeBuffer(buffers.taps, 0, new Float32Array(taps))
      device.queue.writeBuffer(buffers.repSeeds, 0, new Uint32Array(repSeeds))

      const bindGroup = device.createBindGroup({
        layout: pipeline.getBindGroupLayout(0),
        entries: [
          { binding: 0, resource: { buffer: buffers.uniform } },
          { binding: 1, resource: { buffer: buffers.ringA } },
          { binding: 2, resource: { buffer: buffers.ringB } },
          { binding: 3, resource: { buffer: buffers.ys } },
          { binding: 4, resource: { buffer: buffers.stepOut } },
          { binding: 5, resource: { buffer: buffers.scratch } },
          { binding: 6, resource: { buffer: buffers.cum } },
          { binding: 7, resource: { buffer: buffers.taps } },
          { binding: 8, resource: { buffer: buffers.repSeeds } },
        ],
      })

      // Steps per submit, sized to keep one dispatch well under any device
      // timeout while amortizing the submit/await round trip.
      const chunk = Math.max(8, Math.min(512, Math.round(3e7 / (N * (ringLen + 30)))))
      const uniform = new ArrayBuffer(48)
      const u32 = new Uint32Array(uniform)
      const f32 = new Float32Array(uniform)
      u32[0] = N
      u32[1] = ringLen
      u32[2] = n
      f32[8] = h0
      f32[9] = delta / h0
      f32[10] = delta
      f32[11] = Math.log(N)

      for (let startStep = 0; startStep < n; startStep += chunk) {
        u32[3] = startStep
        u32[4] = Math.min(chunk, n - startStep)
        device.queue.writeBuffer(buffers.uniform, 0, uniform)
        const encoder = device.createCommandEncoder()
        const pass = encoder.beginComputePass()
        pass.setPipeline(pipeline)
        pass.setBindGroup(0, bindGroup)
        pass.dispatchWorkgroups(R)
        pass.end()
        device.queue.submit([encoder.finish()])
        await device.queue.onSubmittedWorkDone()
        onProgress?.(Math.min(startStep + chunk, n), n)
      }

      const encoder = device.createCommandEncoder()
      encoder.copyBufferToBuffer(buffers.stepOut, 0, buffers.readback, 0, R * n * 8)
      device.queue.submit([encoder.finish()])
      await buffers.readback.mapAsync(GPUMapMode.READ)
      const steps = new Float32Array(buffers.readback.getMappedRange().slice(0))
      buffers.readback.unmap()

      // Per-step values accumulate in f64 on the CPU, where the lost-lock
      // diagnostic is evaluated exactly as in the scalar filter.
      const limit = collapseThreshold(delta, h0, DEFAULT_COLLAPSE_MARGIN)
      return Array.from({ length: R }, (_, r) => {
        let loglik = 0
        let essSum = 0
        let essMin = 1
        let nLost = 0
        let firstLost = -1
        for (let t = 0; t < n; t++) {
          const step = steps[(r * n + t) * 2]
          const ess = steps[(r * n + t) * 2 + 1]
          loglik += step
          essSum += ess
          essMin = Math.min(essMin, ess)
          if (step < limit) {
            nLost++
            if (firstLost < 0) firstLost = t
          }
        }
        const result: FilterResult = {
          loglikNats: loglik,
          nSteps: n,
          entropyRateBits: -loglik / (n * Math.LN2),
          meanEss: essSum / n,
          minEss: essMin,
          nLostLock: nLost,
          firstLostStep: firstLost,
          lostLockFraction: nLost / n,
          collapseThreshold: limit,
        }
        return result
      })
    } finally {
      for (const b of Object.values(buffers)) b.destroy()
    }
  }
}
